### ISSUE TITLE
Numeric stability issue with full parameter fine tuning of Llama 2 7B model with FSDP

### ISSUE THREAD

**Issue by xiaohai2016** [pytorch/pytorch#120383]
### 🐛 Describe the bug

I was wondering anyone having experiences with full parameter fine tuning of Llama 2 7B model using FSDP can help: I put in all kinds of seeding possible to make training deterministic (https://pytorch.org/docs/stable/notes/randomness.html); however I still observe that the backward gradients on the first sample training vary on each run. The variation of gradients is around the scale of 1.0e-8.

I am using FSDP to wrap around the decoder module.

I don’t have the numeric stability issue if I only fine tune an MLP classification head. The numeric instability seems to occur as soon as the decoder layers are wrapped in FSDP and require gradients.

The numeric instability causes each of my training run to produce models of noticeably different qualities. Any help or suggestions would be appreciated!

### Versions

Collecting environment information...
PyTorch version: 2.1.2+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.11.7 | packaged by conda-forge | (main, Dec 23 2023, 14:43:09) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-5.15.120+-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.3.107
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB
GPU 2: NVIDIA A100-SXM4-40GB
GPU 3: NVIDIA A100-SXM4-40GB

Nvidia driver version: 525.125.06
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          96
On-line CPU(s) list:             0-95
Thread(s) per core:              2
Core(s) per socket:              24
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           85
Model name:                      Intel(R) Xeon(R) CPU @ 2.20GHz
Stepping:                        7
CPU MHz:                         2200.204
BogoMIPS:                        4400.40
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       1.5 MiB
L1i cache:                       1.5 MiB
L2 cache:                        48 MiB
L3 cache:                        77 MiB
NUMA node0 CPU(s):               0-23,48-71
NUMA node1 CPU(s):               24-47,72-95
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:          Mitigation; Enhanced IBRS
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Mitigation; Clear CPU buffers; SMT Host state unknown
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat avx512_vnni md_clear arch_capabilities

Versions of relevant libraries:
[pip3] flake8==7.0.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.3
[pip3] pytorch-lightning==2.1.3
[pip3] pytorch-ranger==0.1.1
[pip3] torch==2.1.2+cu121
[pip3] torch-cluster==1.6.3+pt21cu121
[pip3] torch_geometric==2.4.0
[pip3] torch-optimizer==0.3.0
[pip3] torch-scatter==2.1.2+pt21cu121
[pip3] torch-sparse==0.6.18+pt21cu121
[pip3] torch-spline-conv==1.2.2+pt21cu121
[pip3] torchaudio==2.1.2+cu121
[pip3] torchensemble==0.1.9
[pip3] torchmetrics==1.3.0.post0
[pip3] torchrec-nightly==2023.8.20
[pip3] torchvision==0.16.2+cu121
[pip3] triton==2.1.0
[conda] numpy                     1.26.3          py311h64a7726_0    conda-forge
[conda] pytorch-lightning         2.1.3                    pypi_0    pypi
[conda] pytorch-ranger            0.1.1                    pypi_0    pypi
[conda] torch                     2.1.2+cu121              pypi_0    pypi
[conda] torch-cluster             1.6.3+pt21cu121          pypi_0    pypi
[conda] torch-geometric           2.4.0                    pypi_0    pypi
[conda] torch-optimizer           0.3.0                    pypi_0    pypi
[conda] torch-scatter             2.1.2+pt21cu121          pypi_0    pypi
[conda] torch-sparse              0.6.18+pt21cu121          pypi_0    pypi
[conda] torch-spline-conv         1.2.2+pt21cu121          pypi_0    pypi
[conda] torchaudio                2.1.2+cu121              pypi_0    pypi
[conda] torchensemble             0.1.9                    pypi_0    pypi
[conda] torchmetrics              1.3.0.post0              pypi_0    pypi
[conda] torchrec-nightly          2023.8.20                pypi_0    pypi
[conda] torchvision               0.16.2+cu121             pypi_0    pypi
[conda] triton                    2.1.0                    pypi_0    pypi

**Comment by xiaohai2016** [comment]
Further experiments shows that if I freeze the first 30 layers plus the embedding layer (keep the last layer and the lm_head layer require_gradient=True), the training is stable. That is the gradients on every run are the same. But once I unfroze any more parameters, the training becomes unstable, even if I set the learning rate to 1.0e-8.

**Comment by xiaohai2016** [comment]
Seeding used for every process:

def set_seed(seed: int = 42) -> None:
    np.random.seed(seed)
    random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    # When running on the CuDNN backend, two further options must be set
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False
    # Set a fixed value for the hash seed
    os.environ["PYTHONHASHSEED"] = str(seed)

**Comment by awgu** [comment]
Could you share how you are applying FSDP (e.g. wrapping policy and args)?

It seems possible that NCCL reduction collectives are not deterministic, in which case there could be slight numeric differences from run to run.

**Comment by xiaohai2016** [comment]
@awgu thank you very much for helping.

The wrapping policy is very simple and derived from llama-recipies as:
```
def lm_auto_wrap_policy(model):
    return functools.partial(
        transformer_auto_wrap_policy,
        transformer_layer_cls={LlamaDecoderLayer,}
    )
```

The wrapping codes:
```
    llama2_model_wrapped = FSDP(
        llama2_model,
        auto_wrap_policy=fsdp_auto_wrap_policy(llama2_model, args),
        mixed_precision=bfSixteen if bf16_ready and args.use_bf16 else fpSixteen if args.use_fp16 else None, # fp32_policy
        sharding_strategy=ShardingStrategy.FULL_SHARD,
        device_id=torch.cuda.current_device(),
        sync_module_states=True, # To broadcast weights from rank 0 device
        param_init_fn=lambda module: module.to_empty(device=torch.device("cuda"), recurse=False)
            if local_rank != 0 else None,
    )
```

I was suspecting non-determinism from NCCL as well. However:
1. The suggested NCCL_RINGS environment variable seems to be deprecated in newer versions and it did not help in my case
2. On the inference run, the results are deterministic in my case. The non-determinism only occurs with gradients.

**Comment by xiaohai2016** [comment]
Inside fsdp_auto_wrap_policy, there is an if statement to choose lm_auto_wrap_policy

**Comment by xiaohai2016** [comment]
I made some progress in debugging. Adding the following codes into set_seed helped to stablize fp32 training:
```
    # deterministic algorithms
    os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
    torch.use_deterministic_algorithms(True)

```
But I see that mixed precision training with bfloat16 is still not numerically stable. Any suggestions? @awgu

**Comment by awgu** [comment]
I did not have a chance to investigate whether the gradient reduce-scatter has non-determinism or not (but if so, that would explain the issue).

Taking a step back, if I am understanding correctly, the issue is that your training is not stable, and you want to stabilize it? Or are you specifically interested in producing numerically identical results after each run?

**Comment by xiaohai2016** [comment]
apologize for the confusion. To clarify: my interest is to produce numerically identical results with each run.

Now, I am able to produce numerically identical results with FP32 training, but not yet with bfloat16 (mixed precision with FSDP). Thanks.

**Comment by awgu** [comment]
Maybe you can try fp32 reduce-scatter?
Change
```
mixed_precision=bfSixteen
```
to
```
mixed_precision=MixedPrecision(param_dtype=torch.bfloat16, reduce_dtype=torch.float32)
```
?

(`from torch.distributed.fsdp import MixedPrecision` if needed)

**Comment by xiaohai2016** [comment]
I experimented, and I was able to have full numerical reproducibility with the following mixed precision training. It seems that the issue is not with the reduce-scatter, but with the parameter type.

```
bfSixteen = MixedPrecision(
    param_dtype=torch.float32,
    reduce_dtype=torch.bfloat16,
    buffer_dtype=torch.bfloat16,
)
```

This is **very** undesirable in that setting `param_dtype=torch.float32` instead of `param_dtype=torch.bfloat16` eliminates the speed boost (around 2x) from the mixed precision. Hence there is little reason left for using mixed precision as opposed to using fp32.

**Comment by awgu** [comment]
I wonder if, because bf16 uses the tensor cores, there is non-determinism that cannot be disabled.
